flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java

```diff
@@ -542,6 +573,16 @@ public void onBuffer(Buffer buffer, int sequenceNumber, int backlog) throws IOEx
 			if (backlog >= 0) {
 				onSenderBacklog(backlog);
 			}
+
+			if (checkpointBarrier != null) {
+				receivedCheckpointId = checkpointBarrier.getId();
+			}
+			if (checkpointBarrier != null && inputGate.bufferReceivedListener != null) {
+				inputGate.bufferReceivedListener.notifyBarrierReceived(checkpointBarrier, channelInfo);
+			}
+			if (notifyReceivedBuffer != null && inputGate.bufferReceivedListener != null) {
```

Comment:
I think if `notifyReceivedBuffer != null` but `bufferReceivedListener is null`
then `notifyReceivedBuffer` should be recycled (or retained here, not before).

Reply:
That's a tricky one. `receivedCheckpointId < lastRequestedCheckpointId` can only be true if the checkpoint has been triggered through `Unaligner`, which in turn guarantees that `bufferReceivedListener != null`. 

Aside from your suggestions, we could also add this assumption to the place, where we create `notifyReceivedBuffer` to make it more explicit. (So don't retain if `bufferReceivedListener == null`). Then we would just trust branch prediction to effectively get rid of the check.

Reply:
Yes, it's tricky and also fragile :)

Why not increase ref count when it's needed?
```
if (notifyReceivedBuffer != null && inputGate.bufferReceivedListener != null) {
  inputGate.bufferReceivedListener.notifyBufferReceived(
    notifyReceivedBuffer.retainBuffer(), <-- here
    channelInfo);
}
```

Reply:
Because the buffer may have already been recycled in the meantime - added to the queue and being fully consumed by task thread.

Reply:
> which in turn guarantees that bufferReceivedListener != null.

In that case couldn't we semi self document it via `checkState()`?:
```
if (notifyReceivedBuffer != null) {
	checkState(inputGate.bufferReceivedListener != null);
	inputGate.bufferReceivedListener.notifyBufferReceived(notifyReceivedBuffer, channelInfo);
}
```
?

Reply:
I extracted `hasListener = inputGate.bufferReceivedListener != null` variable and tried to add it a) as a guard around the treatment of `notifyBuffer/Barrier` and b) in checkState. Got a warning that it is always `true` for `notifyBuffer/Barrier != null`. So left it out for now. Feel free to nag once more and I'll add it anyways.